### PR TITLE
Add some VFS helpers

### PR DESCRIPTION
**Pull request by brenns10**
- A corelens module "ls" for listing dentry children.
- A corelens module "fsnotify" for printing fsnotify info about every fanotify, inotify, and dnotify group.
- Some minor improvements to bt() in order to improve the fsnotify helper.

### CODE REVIEW

**Comment by brenns10** [comment]
Sample inotify and dnotify output from my laptop:
```
[PID 4071095 COMM: gvfsd-trash inotify FD 10]
FSNOTIFY GROUP: ffff8bf0b8574000
  MARK: inode ffff8bee62d7b408 /run/mount/utab.event
    CNT:2 MASK:CLOSE_WRITE|UNMOUNT IGN:0
OBJECT SUMMARY: inode: 1
0 notifications are pending.
Tasks waiting for notification:
  [EVENTPOLL: ffff8bee40280b40]
  Waiting in file->poll():
    [PID: 4071095 COMM: gvfsd-trash WAIT: select]

[SYSTEM DNOTIFY GROUP]
FSNOTIFY GROUP: ffff8bee427b5000
OBJECT SUMMARY:
0 notifications are pending.
Tasks waiting for notification:
  <no waiters>
```
Sample fanotify output:
```
[PID 3756 COMM: ds_am fanotify FD 7]
FSNOTIFY GROUP: ffff9bbd820b9d00
  MARK: inode ffff9bc1228d9640 /var/cache/uptrack/Linux/x86_64/5.15.0-206.153.7.el8uek.x86_64/#2 SMP Thu May 9 15:52:29 PDT 2024/results.server-stamp
    CNT:1 MASK:0 IGN:0
  MARK: vfsmount ffff9bbd8b244a00 FS:xfs MOUNT:/
    CNT:76 MASK:CLOSE_WRITE|OPEN_PERM|OPEN_EXEC_PERM IGN:0
   ... note: several lines removed here for brevity ...
OBJECT SUMMARY: inode: 131221, vfsmount: 32
84 notifications are pending.
Tasks waiting for notification:
  <no waiters>
0 pending permission responses
Tasks waiting for permission response from userspace:
  [PID: 1828676 COMM: sshd WAIT: direct]
  ... several lines removed ...
```

With higher levels of verbosity, at level 3, all of the ~80 tasks which were waiting for a permission response would have gotten printed with a stack trace. At level 4, all 131k inode marks would hav ebeen printed (by default, we only print the first 10 inodes, but we print all super blocks and vfsmounts).

Of course, for the "ls" module it's a bit less interesting.

```
>>> cl("ls /")
00000 /__init__.py
00001 /conftest.py
00002 /pytest.ini
00003 /tox.ini
[...]
00439 /sys
00440 /usr
00441 /etc
23 positive, 419 negative dentries
>>> cl("ls -c /")
23 positive, 419 negative dentries
```

The ls module can be quite slow at times due to the implementation of [path_lookup()](https://drgn.readthedocs.io/en/latest/helpers.html#drgn.helpers.linux.fs.path_lookup) which iterates over child lists rather than using the hash table. I'll try to upstream a helper that can use the hash table as well.

**Comment by brenns10** [comment]
There's a bit of overlap in one of my waitqueue functions and the epoll code that @imran-kn has in #83, I'll see if there is anything that can be shared between these branches.

**Review by biger410 — COMMENTED** [review]
One comment, other looks good.

File: drgn_tools/bt.py [hunk 1]
```diff
@@ -570,13 +570,27 @@ def bt_has(
...
     for cpu in for_each_online_cpu(prog):
+        task = cpu_curr(prog, cpu)
+        if skip_idle and task.comm.string_().decode() == f"swapper/{cpu}":
+            # Just because it's the swapper task, does not mean it is idling.
+            # Check the symbol at the top of the stack to ensure it's the
+            # architecture idle function.
+            trace = prog.stack_trace(task)
+            try:
+                sym = trace[0].symbol().name
+                if sym in ("intel_idle",):
```
[the remark this answers is not in the document]
Comment (brenns10) [reply drgn_tools/bt.py:590]: Good catch, no:
```
PID: 0        TASK: ffff0000c0388000 [R] CPU: 1!  COMMAND: "swapper/1"
!# 0 [ffff80000b8cbdd0] arch_cpu_idle at 0xffff80000917c7ac arch/arm64/kernel/idle.c:46:1
 # 1 [ffff80000b8cbde0] default_idle_call at 0xffff80000918a208 kernel/sched/idle.c:112:3
 # 2 [ffff80000b8cbe00] cpuidle_idle_call at 0xffff80000814e1c8 kernel/sched/idle.c:194:3
 # 3 [ffff80000b8cbe40] do_idle at 0xffff80000814e2e0 kernel/sched/idle.c:306:4
 # 4 [ffff80000b8cbe70] cpu_startup_entry at 0xffff80000814e590 kernel/sched/idle.c:403:3
 # 5 [ffff80000b8cbe90] secondary_start_kernel at 0xffff80000806e608 arch/arm64/kernel/smp.c:265:2
 # 6 [ffff80000b8cbeb0] __secondary_switched at 0xffff80000919e30c arch/arm64/kernel/head.S:661:0
```

Adding `arch_cpu_idle` would help for arm. There's also a few more that are safe to ignore. E.g. if the core was halted during idle on x86_64 it looks like this:
```
PID: 0        TASK: ffff8b8007e70000 [R] CPU: 3!  COMMAND: "swapper/3"
!# 0 [ffffa25ac008be70] native_safe_halt at 0xffffffff827fc0e2 ./arch/x86/include/asm/irqflags.h:61:1
 # 1 [ffffa25ac008be80] arch_safe_halt at 0xffffffff827fbcd2 ./arch/x86/include/asm/paravirt.h:144:2
 # 2 [ffffa25ac008be80] default_idle at 0xffffffff827fbcd2 arch/x86/kernel/process.c:579:2
 # 3 [ffffa25ac008beb0] arch_cpu_idle at 0xffffffff81e40e85 arch/x86/kernel/process.c:570:2
 # 4 [ffffa25ac008bec0] default_idle_call at 0xffffffff827fc2cc kernel/sched/idle.c:94:3
 # 5 [ffffa25ac008bed0] cpuidle_idle_call at 0xffffffff81ee54df kernel/sched/idle.c:154:3
 # 6 [ffffa25ac008bed0] do_idle at 0xffffffff81ee54df kernel/sched/idle.c:264:4
 # 7 [ffffa25ac008bf18] cpu_startup_entry at 0xffffffff81ee56ed kernel/sched/idle.c:356:3
 # 8 [ffffa25ac008bf30] start_secondary at 0xffffffff81e6e226 arch/x86/kernel/smpboot.c:265:2
 # 9 [ffffa25ac008bf58] secondary_startup_64 at 0xffffffff81e000e6 arch/x86/kernel/head_64.S:241:0
```

So `native_safe_halt` is another to add. Thankfully, the only consequence of missing functions here is that the `bt -a` output is slightly longer. We can always add more, but we won't accidentally miss important stacks.

**Comment by brenns10** [comment]
Gitlab vmcore & VM tests pass!

**Review by biger410 — APPROVED** [review]
Looks good.

**Pull request merged**